cores/arduino/stm32/dma.c

```diff
@@ -0,0 +1,293 @@
+      
+      /* Set clock as active */
+      channelNumber = pinmap_find_dmaChannelNumber(hadc->Instance, DmaMap_DMA);
+      dmaNumber     = pinmap_find_dma(hadc->Instance, DmaMap_DMA);
+      
+      /* NVIC configuration for DMA Input data interrupt */
+      *irqHandler = pinmap_find_dmaIrqHandler(hadc->Instance, DmaMap_DMA);
+      #ifdef ADC1_BASE
+      if (hadc->Instance == ADC1)
+      {
+        *indice = 0;
+      }
+      #endif
+      #ifdef ADC2_BASE
+      if (hadc->Instance == ADC2)
+      {
+         *indice = 1;
+      }
+      #endif
+      #ifdef ADC3_BASE
+      if (hadc->Instance == ADC3)
+      {
+        *indice = 2;
+      }
+      #endif
+      #ifdef ADC4_BASE
+      if (hadc->Instance == ADC4)
+      {
+        *indice = 3;
+      }
+      #endif
+    };break;
+    /*case SPI :
+    {
+      spi = (SPI_HandleTypeDef*)periph_handle;
+    };break;*/ //to do
+  }
+  
+#ifdef __HAL_RCC_DMA1_CLK_ENABLE
+  __HAL_RCC_DMA1_CLK_ENABLE();
+#endif
+
+#ifdef __HAL_RCC_DMA2_CLK_ENABLE
+  __HAL_RCC_DMA2_CLK_ENABLE();
+#endif
+  
+  /* Set clock as active */
+  dmaTab[dmaNumber-1][channelNumber] = true;
+}
+
+/**
+  * @brief  This function will deinit the dma and channel corresponding 
+  *         to the wanted peripheral
+  * @param  periph_handle : pointer to the peripheral HandleTyeDef
+  * @param  periph        : the peripheral type to use
+  * @retval none
+  */
+void dma_free(void *periph_handle, Peripheral periph)
+{
+  IRQn_Type irqHandler    = 0;
+  int8_t    channelNumber = 0;
+  int8_t    dmaNumber     = 0;
+  ADC_HandleTypeDef *hadc = NULL;
+  //SPI_HandleTypeDef *spi  = NULL; to do
+  
+  hadc = (ADC_HandleTypeDef*)periph_handle;
+  
+  switch(periph)
+  {
+    case ADC_PERIPH :
+    {
+      irqHandler = pinmap_find_dmaIrqHandler(hadc->Instance, DmaMap_DMA);
+      HAL_DMA_DeInit(hadc->DMA_Handle);
+      channelNumber = pinmap_find_dmaChannelNumber(hadc->Instance, DmaMap_DMA);
+      dmaNumber     = pinmap_find_dma(hadc->Instance, DmaMap_DMA);
+    }; break;
+    
+    /*case SPI :
+    {
+      
+    }; break;*/ //to do
+  }
+  
+  HAL_NVIC_DisableIRQ(irqHandler);
+  
+  // Disable the clock only if not another channel of the dma is use
+  dmaTab[dmaNumber-1][channelNumber] = false;
+  
+ for (int i=0; i<CHANNEL_NUMBER; i++)
+  {
+    if (dmaTab[dmaNumber][i] == true)
+    {
+      break;
+    }
+    if (i == CHANNEL_NUMBER-1)
+    {
+      // All dma channel are inactive
+      if (dmaNumber == 1)
+      {
+      #ifdef __HAL_RCC_DMA1_CLK_ENABLE
+        __HAL_RCC_DMA1_CLK_DISABLE();
+      #endif
+      }
+      else if (dmaNumber == 2)
+      {
+        #ifdef __HAL_RCC_DMA2_CLK_ENABLE
+        __HAL_RCC_DMA2_CLK_DISABLE();
+        #endif
+      }
+    }
+  }
+}
+
+
+/**
+  * @brief  DMA IRQ Handle 
+  * @param  None
+  * @retval None
+  */
+
+#ifdef ADC1_BASE
+void DMA2_Stream0_IRQHandler(void)
+{
+  #if defined(STM32F2xx) || defined(STM32F4xx) || defined(STM32F7xx)
+  if (AdcHandleTab[0].DMA_Handle != NULL)
```

Comment:
from dma point of view this is a HandleTab, we don't need to to know it's a Adc  one.

Reply:
Where is the DMA_Handle table initiliazed with proper DMA_handle values ?

Reply:
This is done by this function, when your link the DMA to a peripheral : 
__HAL_LINKDMA(hadc, DMA_Handle, DmaHandleTab[indice]);

Reply:
Ok same as previous - I have lloked at __HAL_LINKDMA and this is now clearer.